game-core/src/main/java/games/strategy/engine/data/util/BreadthFirstSearch.java

```diff
@@ -0,0 +1,91 @@
+package games.strategy.engine.data.util;
+
+import games.strategy.engine.data.GameMap;
+import games.strategy.engine.data.Territory;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.ObjectUtils;
+
+/**
+ * Implements Breadth First Search (BFS) to traverse / find territories. Since the search criteria
+ * varies depending on the use case, the class is designed to be sub-classed, with methods visit()
+ * and shouldContinueSearch() that can be overridden to customize the behavior.
+ */
+public class BreadthFirstSearch {
+  private final GameMap map;
+  @Getter private final Set<Territory> visited;
+  private final ArrayDeque<Territory> territoriesToCheck;
+  private final Predicate<Territory> cond;
+
+  public BreadthFirstSearch(final Territory startTerritory, final Predicate<Territory> cond) {
+    this.map = startTerritory.getData().getMap();
+    this.visited = new HashSet<Territory>(List.of(startTerritory));
+    this.territoriesToCheck = new ArrayDeque<Territory>(List.of(startTerritory));
+    this.cond = cond;
+  }
+
+  public BreadthFirstSearch(final Territory startTerritory) {
+    this(startTerritory, Matches.always());
+  }
+
+  /**
+   * Called when a new territory is encountered. Can be overridden to provide custom search
+   * behavior.
+   *
+   * @param territory The new territory.
+   */
+  public void visit(final Territory territory) {}
```

Comment:
Can we make this method abstract? Overriding concrete methods can get pretty complex in a hurry. Another strategy that would avoid inheritance completely would be to pass in a constructor object that defines this behavior. EG:

```
private final Consumer<Territory> visitor;

  public void visit(final Territory territory) {
      visitor.accept(terrritory);
}
````

Looking at the other methods, it looks like we might be providing a lot of default behavior but any/all of it can be overriden. If this behavior were injected it could be more deterministic and easier to follow.  Essentially you would know have to know about instance types and the states of objects to know how they would function, WDYT?

Reply:
Apologies @asvitkine if this is getting pedantic for you, just want to be sure the suggestion for how to avoid overriding concrete methods is somewhat clear so that you can respond or incorporate it.

Example:

```
class Parent {
  int computeSomething() {
       return 1;
   }
}

@Override
class Child extends Parent{
  int computeSomething() {
       return 10;
   }
}
```

In the above, there are really two behaviors, we can expand the inheritance tree to flatten the structure, ie instead of:
```
  P
    \
     C(1)
```

We get a graph that looks like this:
```
       P
     /   \
 C(1)   C(2)
```

That would look like this:

```
absract class Parent {
  abstract int computeSomething();
}

@Override
class Child1 extends Parent{
  int computeSomething() {
       return 1;
   }
}

@Override
class Child2 extends Parent{
  int computeSomething() {
       return 10;
   }
}
```

The other recommendation is to inject behavior via configuration, in which case there is no inheritance at all, eg:

```

@AllArgsBuilder
class Parent {
  private final Supplier<Integer> computation;
  int computeSomething() {
     return computation.get();
  }
}


class SomeClass {
   Parent child1 = new Parent(() -> 1);
   Parent child2 = new Parent(() -> 10);
}
```

Reply:
We could go with an interface or abstract class that has the two methods to override that are passed from the caller.  I agree that it tightens things up a bit more (i.e. the subclasses of that abstract can't do as much), so I can look at doing that if you prefer. 

Though it doesn't change things too much - the caller now just declares an anonymous class that extends the new abstract class/interface that has just the visit() and shouldContinueSearch() methods. Let me know if you prefer that approach.

My original thought was to keep it simpler to start with, so that as we apply this to more cases (like getNeighbours(distance) suggestion and possibly getRoute()), we can figure out what the API surface would need to be like to support them and evolve as necessary.

Reply:
> Though it doesn't change things too much - the caller now just declares an anonymous class that extends the new abstract class/interface that has just the visit() and shouldContinueSearch() methods rather than extending BFS. Let me know if you prefer that approach.

Assuming it can be done cleanly, I think that sounds good/preferable :+1: 

It sounds like that would avoid a concrete implementation from having an override. Such scenarios, concrete implementations having an override is usually where inheritance starts to become quite bad.

Reply:
Done.